presto-hive/src/main/java/io/prestosql/plugin/hive/util/SequenceFileRecordWriter.java

```diff
@@ -64,11 +64,34 @@ public void write(Writable writable)
     public void close(boolean abort)
             throws IOException
     {
-        try (Closeable ignored = writer) {
```

Comment:
Thanks for implementing try-with-resources manually. I know it's annoying to write. However, I think we can simplify this using `Closer` from Guava, which avoids self suppression
```java
// use Closer to avoid self-suppression
try (Closer closer = Closer.create()) {
    closer.register(writer);

Reply:
@electrum that's much cleaner, will add that

Reply:
I think I am missing something:
In what way is
```java
try (Closeable ignored = writer) {
   ...
```
semantically different from
```java
try (Closer closer = Closer.create()) {
   closer.register(writer);
   ...
```

In both cases try-with-resources is in place and if exception is thrown from `try` block the possible exception thrown from `writer.close()` will be suppressed.

Reply:
Both options presented here (try-with-resources directly, or with closer) are equivalent from "Self-suppression not permitted" perspective (both end up throwing the "Self-suppression not permitted" instead of actual).

```
public static void main(String[] args)
        throws Exception
{
//        try (Gizmo x = new Gizmo()) {
//            x.thr();
//        }

    try (Closer closer = Closer.create()) {
        Gizmo x = closer.register(new Gizmo());
        x.thr();
    }
}

private static class Gizmo
        implements Closeable
{
    private final RuntimeException e = new RuntimeException("gizmo!!!");

    public void thr()
    {
        throw e;
    }

    @Override
    public void close()
    {
        throw e;
    }
}
```

The way out is using `Closer` the Java 7 way

```java
Closer closer = Closer.create();
Gizmo x = closer.register(new Gizmo());
try {
    x.thr();
}
catch (Exception e) {
    throw closer.rethrow(e);
}
finally {
    closer.close();
}
```

while verbose, this is still simpler than implementing try-with-resource manually, with correct suppression

```
Throwable thrown = null;
Gizmo x = new Gizmo();
try {
    x.thr();
}
catch (Throwable e) {
    thrown = e;
}
finally {
    try {
        x.close();
    }
    catch (Throwable closeException) {
        if (thrown == null) {
            thrown = closeException;
        }
        else if (thrown != closeException) {
            thrown.addSuppressed(closeException);
        }
    }
}
if (thrown != null) {
    propagateIfPossible(thrown, IOException.class);
    throw new RuntimeException(thrown);
}
```